### PR TITLE
feat: protect segment operations for change requests

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Update segment and delete segment now have protected mode which means that with CR enabled you can't change them willy-nilly. You either need skip permission for any project environment or go through a change request.

There's a corresponding enterprise change where I have tests and fix breaking changes.

## Discussion points

I noticed that we have interface for the segment service. Since this is the only service with an interface it looks odd. I'm tempted to remove it since there's only one impl of the interface and TS is structurally typed. @gastonfournier since you added the interface WDYT?

### CODE REVIEW

File: src/test/fixtures/fake-project-store.ts [hunk 1]
```diff
@@ -177,7 +177,7 @@ export default class FakeProjectStore implements IProjectStore {
         projectId: string,
         // eslint-disable-next-line @typescript-eslint/no-unused-vars
         environment: string,
-    ): Promise<CreateFeatureStrategySchema | undefined> {
+    ): Promise<CreateFeatureStrategySchema | null> {
```
Comment (kwasniew) [inline src/test/fixtures/fake-project-store.ts:180]: TS error fix

File: src/lib/services/segment-service.ts [hunk 2]
```diff
@@ -108,13 +130,26 @@ export class SegmentService implements ISegmentService {
 
         await this.eventStore.store({
             type: SEGMENT_UPDATED,
-            createdBy: user.email || user.username,
+            createdBy: user.email || user.username || 'unknown',
```
Comment (kwasniew) [inline src/lib/services/segment-service.ts:133]: TS error fix

File: src/lib/services/segment-service.ts [hunk 3]
```diff
@@ -82,17 +96,25 @@ export class SegmentService implements ISegmentService {
...
             data: segment,
         });
 
         return segment;
     }
 
-    async update(
+    async update(id: number, data: unknown, user: User): Promise<void> {
+        if (this.flagResolver.isEnabled('segmentChangeRequests')) {
```
Comment (kwasniew) [inline src/lib/services/segment-service.ts:107]: we have to be careful not to enable it before the UI is ready

File: src/lib/features/change-request-access-service/sql-change-request-access-read-model.ts [hunk 4]
```diff
@@ -35,6 +35,23 @@ export class ChangeRequestAccessReadModel
         return !(changeRequestEnabled && !canSkipChangeRequest);
     }
 
+    public async canBypassChangeRequestForProject(
```
Comment (kwasniew) [inline src/lib/features/change-request-access-service/sql-change-request-access-read-model.ts:38]: no check for environment. If a user can skip change requests in any env we're good for now. We can tweak this logic later. For now need something cheap to experiment

File: src/lib/features/change-request-access-service/sql-change-request-access-read-model.ts [hunk 5]
```diff
@@ -35,6 +35,23 @@ export class ChangeRequestAccessReadModel
...
+            user
+                ? this.accessService.hasPermission(
+                      user,
+                      SKIP_CHANGE_REQUEST,
+                      project,
+                  )
+                : Promise.resolve(false),
+            this.isChangeRequestsEnabledForProject(project),
+        ]);
+        return !(changeRequestEnabled && !canSkipChangeRequest);
```
Comment (thomasheartman) [inline src/lib/features/change-request-access-service/sql-change-request-access-read-model.ts:52]: Could we express this without the double negation? I'm struggling to parse what we're saying here.

File: src/lib/services/segment-service.ts (hunk 3 above)
Comment (thomasheartman) [reply src/lib/services/segment-service.ts:107]: Fair. It's currently enabled in dev, but we can turn that off

**Review by thomasheartman — APPROVED** [review]
Nice! My only comment is the double negation. It'd be nice if we could express that a bit clearer, but I'll leave it up to you.

File: src/lib/features/change-request-access-service/sql-change-request-access-read-model.ts (hunk 5 above)
Comment (kwasniew) [reply src/lib/features/change-request-access-service/sql-change-request-access-read-model.ts]: great catch, changing it to  `return canSkipChangeRequest || !changeRequestEnabled;`

**Comment by gastonfournier** [comment]
> I noticed that we have interface for the segment service. Since this is the only service with an interface it looks odd. I'm tempted to remove it since there's only one impl of the interface and TS is structurally typed. @gastonfournier since you added the interface WDYT?

Looking at https://github.com/Unleash/unleash/pull/3315 I think the plan was to use the interface to establish the contract as enterprise and  OSS had different implementations. I don't remember the context but we were moving things that were on OSS to the enterprise repo and this was probably in the expansion phase of the migration and it just stayed... If it doesn't make more sense and you want to remove it, I vote we follow our conventions (and this is out of our code conventions)

**Pull request merged**